casdoorsdk/product.go

```diff
@@ -124,23 +124,28 @@ func (c *Client) DeleteProduct(product *Product) (bool, error) {
 	return affected, err
 }
 
-func (c *Client) BuyProduct(name string, providerName string) (*Product, error) {
+func (c *Client) BuyProduct(name string, providerName string, userName string) (*Product, error) {
 	queryMap := map[string]string{
 		"id":           fmt.Sprintf("%s/%s", c.OrganizationName, name),
 		"providerName": providerName,
+		"userName":     userName,
 	}
 
-	url := c.GetUrl("buy-product", queryMap)
+	bytes, err := c.DoPost("buy-product", queryMap, []byte(""), false, false)
+	if err != nil {
+		return nil, err
+	}
 
-	bytes, err := c.DoGetBytes(url)
+	productJson, err := json.Marshal(bytes.Data)
 	if err != nil {
 		return nil, err
 	}
 
-	var product *Product
-	err = json.Unmarshal(bytes, &product)
+	var product Product
+	err = json.Unmarshal(productJson, &product)
```

Comment:
Why Marshal and Unmarshal again? Refer to existing code

Including var names

Reply:
we can't transfer map object (resp.Data) to Product object without add extra library